packages/cosmic-proto/test/address-hooks.test.js

```diff
@@ -0,0 +1,214 @@
+  title(providedTitle = '', prefix, baseAddress, hookData, charLimit, throws) {
+    let sep = providedTitle.endsWith(' ') ? '' : ' ';
+    const limitDesc = charLimit ? `${sep}charLimit=${charLimit}` : '';
+    if (limitDesc) sep = ' ';
+    const throwsDesc = throws ? `${sep}throws` : '';
+    if (throwsDesc) sep = ' ';
+    return `${providedTitle}${limitDesc}${throwsDesc}`;
+  },
+  exec(t, prefix, addrBytes, hookData, charLimit, throws) {
+    const { encodeBech32, joinHookedAddress, splitHookedAddress } =
+      t.context.addressHooks;
+    const baseAddress = encodeBech32(prefix, addrBytes, charLimit);
+    const make = () => joinHookedAddress(baseAddress, hookData, charLimit);
+    if (throws) {
+      t.throws(make, throws);
+      return;
+    }
+    const encoded = make();
+    t.log('encoded', encoded, addrBytes);
+    const decoded = splitHookedAddress(encoded, charLimit);
+    t.deepEqual(decoded, {
+      baseAddress,
+      hookData,
+    });
+  },
+});
+
+{
+  const charLimit = 90;
+  const prefix = 'agoric';
+  const addrBytes = new Uint8Array(20);
+  for (let j = 0; j < addrBytes.length; j += 1) {
+    addrBytes[j] = j;
+  }
+
+  for (let i = 0; i <= 30; i += 1) {
+    const hookData = new Uint8Array(i);
+    for (let j = 0; j < i; j += 1) {
+      hookData[j] = i - j;
+    }
+    test(
+      `${addrBytes.length}-byte baseAddress, ${i}-byte hookData`,
+      lengthCheckMacro,
+      prefix,
+      addrBytes,
+      hookData,
+      charLimit,
+      i > 23 ? { message: /Exceeds length limit/ } : undefined,
+    );
+  }
+}
+
+/**
+ * @type {import('ava').Macro<
+ *   [
+ *     baseAddress: string,
+ *     query: import('../src/address-hooks.js').HookQuery,
+ *     expected: string,
+ *   ]
+ * >}
+ */
+const addressHookMacro = test.macro({
+  title(providedTitle = '', baseAddress, query) {
+    return `${providedTitle} ${baseAddress} ${JSON.stringify(query)}`;
+  },
+  exec(t, baseAddress, query, expected) {
+    const { encodeAddressHook, splitHookedAddress, decodeAddressHook } =
+      t.context.addressHooks;
+    const encoded = encodeAddressHook(baseAddress, query);
+    t.log('encoded', encoded);
+    t.is(encoded, expected);
+
+    const { baseAddress: ba1, hookData } = splitHookedAddress(encoded);
+    t.is(ba1, baseAddress);
+
+    const td = new TextDecoder();
+    t.log('splitHookedAddress', ba1 + td.decode(hookData));
+
+    const { baseAddress: decodedBaseAddress, query: decodedQuery } =
+      decodeAddressHook(encoded);
+    t.is(decodedBaseAddress, baseAddress);
+    t.deepEqual(decodedQuery, query);
+  },
+});
+
+test(
+  'agoric hook',
+  addressHookMacro,
+  'agoric1qqp0e5ys',
+  { d: null, a: 'b', c: ['d', 'd2'] },
+  'agoric10rchqqplvy7kyfnr84jzvceavsezveqqqyu2w5lp',
+);
+
+test(
+  'cosmos hook',
+  addressHookMacro,
+  'cosmos1qqxuevtt',
+  {
+    everything: null,
+    dst: ['a', 'b', 'c'],
+  },
+  'cosmos10rchqqplv3ehg0tpyej8xapavgnxgum5843jvetkv4e8jargd9hxwqqp4vx73n',
+);
```

Comment:
To check default length limits against expected fast-usdc usage, let's please add a test to replace what's currently in the fast usdc test fixture: `agoric16kv2g7snfc4q24vg3pjdlnnqgngtjpwtetd2h689nz09lcklvh5s8u37ek?EUD=osmo183dejcnmkka5dzcu9xw6mywq0p2m5peks28men`

 - base address is `agoric16kv2g7snfc4q24vg3pjdlnnqgngtjpwtetd2h689nz09lcklvh5s8u37ek`
 - query is `{ EUD: 'osmo183dejcnmkka5dzcu9xw6mywq0p2m5peks28men' }`

Reply:
Encoded as `agoric10rchp4vc53apxn32q42c3zryml8xq3xshyzuhjk6405wtxy7tl3d7e0f8az423padaek6me38qekget2vdhx66mtvy6kg7nrw5uhsaekd4uhwufswqex6dtsv44hxv3cd4jkuqpqvduyhf`

Length: 149

We're mostly paying the 8/5 bech32 encoding tax on every byte of the query string.  If we really need a smaller string, we'd have to resort to adding some compression.